src/main/java/teammates/logic/core/Emails.java

```diff
@@ -661,21 +674,63 @@ public void addEmailToTaskQueue(MimeMessage message, long emailDelayTimer) throw
         
     }
 
-    public void sendEmail(MimeMessage message) throws MessagingException {
-        log.info(getEmailInfo(message));
-        Transport.send(message);        
+    public void sendEmail(MimeMessage message) throws MessagingException, JSONException, IOException {
+        if (Config.isUsingSendgrid()) {
+            Sendgrid email = parseMimeMessageToSendgrid(message);
+            log.info(getEmailInfo(email));
+            
+            try {               
+                email.send();
+            } catch (Exception e) {
+                log.severe("Sendgrid failed, sending with GAE mail");
+                Transport.send(message);  
+            }          
+        } else {
+            log.info(getEmailInfo(message));
+            Transport.send(message);     
+        }       
```

Comment:
If you extract two utility methods `sendUsingSendgrid` and `sendUsingGae`, the logic of this method will become more self explanatory.